### PR TITLE
Fix for Issue 4014: simplified ping handling / deferral based on rece…

### PR DESCRIPTION
**Pull request by sandykellagher**
Fix for issue #4014: using simplified ping handling / deferral based on receiving message or ping from the remote end
This is my own original work that I license to the project

 - [ x] Link to issue, e.g. `Resolves #NNN`
 - [ ] Documentation added (if applicable)
 - [ ] Tests added
 - [ x] Branch rebased on top of current main (`git pull --rebase origin main`)
 - [ x] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [ ] Build is green in Travis CI
 - [ x] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

Resolves #

### Changes proposed in this pull request:

 - add a lastRxMsg to the pinfo structure, and rename last to lastRxPing
 - update lastRxMsg in main read loop when any data message or subscription has been received
 - in pingTimer, hold off sending a ping based on lastRxMsg and lastRxPing values, and remove special coding for different types of connection

### CODE REVIEW

**Comment by sandykellagher** [comment]
I see this change resulted in a couple of JetStream cluster or supercluster tests failing. But it's not clear to me why that should be...

File: server/client.go [hunk 1]
```diff
@@ -286,9 +286,10 @@ type rrTracking struct {
 
 // Struct for PING initiation from the server.
 type pinfo struct {
-	tmr  *time.Timer
-	last time.Time
-	out  int
+	tmr        *time.Timer
+	lastRxPing time.Time
```
Comment (derekcollison) [inline server/client.go:290]: Can these not be combined? I think they are doing the same which is recording inbound activity only.

File: server/client.go [hunk 2]
```diff
@@ -4555,32 +4557,21 @@ func (c *client) processPingTimer() {
...
-
 	pingInterval := c.srv.getOpts().PingInterval
 	if c.kind == GATEWAY {
 		pingInterval = adjustPingIntervalForGateway(pingInterval)
 	}
 	now := time.Now()
 	needRTT := c.rtt == 0 || now.Sub(c.rttStart) > DEFAULT_RTT_MEASUREMENT_INTERVAL
 
-	// Do not delay PINGs for GATEWAY or spoke LEAF connections.
-	if c.kind == GATEWAY || c.isSpokeLeafNode() {
```
Comment (derekcollison) [inline server/client.go:4568]: Doing the write is most important over potential long haul or multi-hope networks. So even if we have received something we may need to write, so prefer this logic remains.

**Review by derekcollison — COMMENTED** [review]
In general looks good, a few minor comments. Thanks!

**Comment by sandykellagher** [comment]
Closing this PR in favour of #4016

**Pull request closed without merging**